chat-android/src/main/java/com/ably/chat/Utils.kt

```diff
@@ -217,3 +226,60 @@ internal class LatestJobExecutor {
         }
     }
 }
+
+/**
+ * A custom implementation using Channel that supports optional awaitable emissions.
+ * This class allows emitting values with optional completion tracking.
+ */
+internal class AwaitableChannel<T>(private val logger: Logger) {
+    private val channel = Channel<Pair<T, CompletableDeferred<Unit>>>(Channel.UNLIMITED)
+    private val completionDeferred = AtomicReference<CompletableDeferred<Unit>?>(null)
+    private var activeCollector = false
+
+    /**
+     * Sends a value to the channel. If `awaitable` is true, the emission is tracked
+     * using a CompletableDeferred to allow awaiting its completion.
+     *
+     * @param value The value to send.
+     */
+    fun trySend(value: T) {
```

Comment:
if we use `AtomicReference` and expect this method to be thread-safe, we should somehow ensure that `completionDeferred.set(deferred); channel.trySend(value to deferred)` happens atomically. But why can't we simply return the deferred from the send method? (Also, `trySend` isn't a very good name here because we always send, as we have an unlimited buffer)


```kotlin
internal class AwaitableChannel<T>(private val logger: Logger) {
   fun send(value: T): Unit {...}
   fun sendWithCompletion(value: T): CompletableDeferred<Unit> {...}
}
```

and no channel-level `await` method

Reply:
Yeah, actually in our case, we are calling trySend in an event listener, so next trySend call will start only when previous one finishes, so we should be okay with atomic reference. 
I guess we can rename it to `send`